ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/process/search_tree/Tree.java

```diff
@@ -63,20 +67,21 @@ private Tree() {
 
         while (doNewIteration(searchTreeRaoParameters, hasImproved, optimalLeaf.getCost(crac), depth)) {
             Set<NetworkAction> availableNetworkActions = crac.getNetworkActions(network, crac.getPreventiveState(), UsageMethod.AVAILABLE);
-            List<Leaf> generatedLeaves = optimalLeaf.bloom(availableNetworkActions);
+            final List<Leaf> generatedLeaves = optimalLeaf.bloom(availableNetworkActions);
 
             if (generatedLeaves.isEmpty()) {
                 break;
             }
 
-            //TODO: manage parallel computation
-            generatedLeaves.forEach(leaf -> leaf.evaluate(network, crac, referenceNetworkVariant, parameters));
-            generatedLeaves = generatedLeaves.stream().filter(leaf -> leaf.getStatus() == Leaf.Status.EVALUATION_SUCCESS).collect(Collectors.toList());
+            evaluateLeaves(network, crac, referenceNetworkVariant, parameters, generatedLeaves);
+
+            List<Leaf> successfulLeaves = generatedLeaves.stream().filter(leaf -> leaf.getStatus() == Leaf.Status.EVALUATION_SUCCESS).collect(Collectors.toList());
 
             hasImproved = false;
-            for (Leaf currentLeaf: generatedLeaves) {
+            for (Leaf currentLeaf: successfulLeaves) {
                 if (improvedEnough(optimalLeaf.getCost(crac), currentLeaf.getCost(crac), searchTreeRaoParameters)) {
                     hasImproved = true;
+                    optimalLeaf.deletePostOptimResultVariant(crac);
                     depth = depth + 1;
                     optimalLeaf.deletePostOptimResultVariant(crac);
```

Comment:
You do this twice here, it doesn't seem necessary.
optimalLeaf.deletePostOptimResultVariant(crac);

Reply:
indeed